README.md

```diff
@@ -131,15 +144,379 @@ class Admin < Dill::Role
+
+
+## Widget Declaration
+A widget can be declared with the `widget` macro, or as a class.
+
+Macro widget declarations take a CSS or XPath selector as the first argument.
+
+```ruby
+widget :todo_item, '.todo-item'
+widget :todo_item, [:xpath, '//some/node']
+```
+
+Class widget declarations define the widget with selector via the `root` method.
+
+```ruby
+class TodoItem < Dill::Widget
+  root '.todo-item'
+end
+
+class TodoItem < Dill::Widget
+  root :xpath, '//some/node'
+end
+```
+
+
+## Widgets with Procs
+Widgets can take a Proc in addition to a CSS selector or XPath expression, allowing us to more precisely define the widget at call time.
+
+```ruby
+class TodoManager < Dill::Role
+  widget :todo_item, -> (text) { ['.todo-item', text: text] }
+
+  def select_item(description)
+    click :todo_item, description
+  end
+end
+```
+
+
+## Widgets Wait to Find an Element
+The `widget` method blocks, meaning it pauses Ruby code execution, and checks the page for that element until found or it reaches Capybara's timeout limit. It does not pause JavaScript code execution. This is handy for dynamic UI tests.
+
+```ruby
+roles.myrole.create_todo_item("Buy Milk")
+# the item form is submitted to server and on response
+# from the server, a new item is appended to the list.
+# 'widget' waits until the item appears on the list,
+# or Capybara's timeout limit is reached
+
+expect(roles.myrole).to see :todo_item, "Buy Milk"
+```
+
+
+## Widget Root
+The **root** of a widget is the Capybara element itself that Dill abstracts. When a widget is declared with the widget macro, the root declaration is implicit and equal to the element with the css class in the definition.
+
+```ruby
+class TodoManager < Dill::Role
+  widget :todo_item, '.todo-item' do
+    def delete
+      root.find('a.delete').click
+    end
+  end
+end
+```
+
+
+## HTML Attributes
+You can access the id and classes of the widget as well with Dill methods. Other attributes can be accessed from the Capybara element, via the `root` method.
+
+```ruby
+# <a href="/items/1" id="todo_item" class="todo-item right-aligned">
+widget(:todo_item).id #=> "todo_form"
+widget(:todo_item).classes #=> ["todo-item", "right-aligned"]
+widget(:todo_item).root['href'] #=> "/items/1"
+```
+
+
+## Getting All the Widgets
+You can get a list of all the elements that match your selector(s) on the page with the `widgets` method.
+
+```ruby
+# note: `widgets` does not wait like `widget` does
+widgets(:todo_item)
+```
+
+
+## Nested Widgets
+Widgets can be nested inside other widgets. Definining attributes of the inner widget, such as classes, will be scoped to within the outer widget.
+
+```ruby
+widget :todo_item, '.todo-item' do
+  widget :delete_link, 'a.delete'
+end
+```
+
+And can be called from that widget.
+
+```ruby
+widget(:todo_item).click :delete_button
```

Comment:
It may make sense to mention here that `click` doesn't wait inside widgets. It may also make sense to have it wait, when called in this context.